File: OS_kernel/py/07_compliance_healing.py
"""
============================================================
KERNEL AGENTIX OS — 07_compliance_healing.py
Module 7: Compliance & Self-Healing
Kepatuhan Peran · Isolasi Anomali · Imunitas Sistem
============================================================
Depends on: corekernel.py
"""
from corekernel import AgentixKernel, Semantic


class ComplianceHealing(AgentixKernel):
    def __init__(self, config=None):
        super().__init__(config)
        self._contracts  = {}
        self._health     = {}
        self._immunity   = 1.0
        self._anomaly_log = []

    def register_contract(self, agent_id, contract):
        """contract: { 'speed': [0, 1], 'energy': [0.2, 1], ... }"""
        self._contracts[agent_id] = dict(contract)
        self._health[agent_id]    = 1.0

    def compliance_score(self, agent_id, context):
        contract = self._contracts.get(agent_id)
        if not contract: return 1.0
        scores = []
        for field, (lo, hi) in contract.items():
            val    = context.get(field, 0.0)
            rng    = hi - lo + 1e-12
            excess = max(0, lo - val) + max(0, val - hi)
            scores.append(max(0.0, 1.0 - excess / rng))
        return sum(scores) / len(scores) if scores else 1.0

    def update_health(self, agent_id, context, decay_good=0.05, decay_bad=0.2):
        compliance = self.compliance_score(agent_id, context)
        current    = self._health.get(agent_id, 1.0)
        is_good    = int(compliance > 0.5)
        decay      = is_good * decay_good + (1 - is_good) * decay_bad
        new_health = max(0.0, min(1.0, current + decay * (compliance - current)))
        self._health[agent_id] = new_health
        if new_health < 0.3:
            self._anomaly_log.insert(0, {
                'tick': self.iteration, 'agent_id': agent_id,
                'health': new_health, 'compliance': compliance,
            })
        return new_health

    def get_health(self, agent_id):
        return self._health.get(agent_id, 1.0)

    def authority_mask(self, agent_id, threshold=0.3, sharpness=10):
        return Semantic.soft_thresh(self.get_health(agent_id), threshold, sharpness)

    def is_quarantined(self, agent_id, threshold=0.15):
        return int(self.get_health(agent_id) < threshold)

    def update_immunity(self):
        vals = list(self._health.values())
        self._immunity = sum(vals) / len(vals) if vals else 1.0
        return self._immunity

    def get_immunity(self): return self._immunity

    def recovery_pulse(self, boost=0.05):
        bonus = self._immunity * boost
        for aid in self._health:
            self._health[aid] = min(1.0, self._health[aid] + bonus)

    def get_anomaly_log(self, last=10):
        return self._anomaly_log[:last]

    def clear_anomaly_log(self):
        self._anomaly_log = []

File: OS_kernel/js/07_compliance_healing.js
/**
 * ============================================================
 * KERNEL AGENTIX OS — 07_compliance_healing.js
 * Module 7: Compliance & Self-Healing
 * Kepatuhan Peran · Isolasi Anomali · Imunitas Sistem
 * ============================================================
 * Depends on: corekernel.js
 *
 * Compliance: each agent has a "role contract" — a set of
 *   expected valence ranges. Deviation triggers a penalty.
 * Self-Healing: anomalies are soft-isolated via tension damping.
 *   No hard kill switches — anomalous agents lose authority.
 * Immunity: system-level health score gates collective actions.
 */
import { AgentixKernel, Semantic } from './corekernel.js';

class ComplianceHealing extends AgentixKernel {
    constructor(config) {
        super(config);
        this._contracts  = {};   // { agentId: { field: [min, max], ... } }
        this._health     = {};   // { agentId: healthScore ∈ [0,1] }
        this._immunity   = 1.0; // system-wide immunity ∈ [0,1]
        this._anomalyLog = [];   // log of anomaly events
    }

    // ─────────────────────────────────────────
    // ROLE CONTRACT — define expected ranges
    // Fields are context keys; [min, max] is the valid valence band
    // ─────────────────────────────────────────
    registerContract(agentId, contract) {
        // contract: { speed: [0, 1], energy: [0.2, 1], ... }
        this._contracts[agentId] = { ...contract };
        this._health[agentId]    = 1.0;
    }

    // ─────────────────────────────────────────
    // COMPLIANCE SCORE
    // 1.0 = fully compliant, 0.0 = fully anomalous
    // Continuous, no if-then
    // ─────────────────────────────────────────
    complianceScore(agentId, context) {
        const contract = this._contracts[agentId];
        if (!contract) return 1.0;

        const fields = Object.entries(contract);
        if (fields.length === 0) return 1.0;

        const scores = fields.map(([field, [min, max]]) => {
            const val    = context[field] || 0;
            const range  = max - min + 1e-12;
            // Distance from valid band → 0 = in-band, >0 = out
            const excess = Math.max(0, min - val) + Math.max(0, val - max);
            return Math.max(0, 1 - excess / range);
        });

        return scores.reduce((a, b) => a + b, 0) / scores.length;
    }

    // ─────────────────────────────────────────
    // HEALTH UPDATE — exponential moving average
    // Integrates compliance over time
    // ─────────────────────────────────────────
    updateHealth(agentId, context, decayGood = 0.05, decayBad = 0.2) {
        const compliance = this.complianceScore(agentId, context);
        const current    = this._health[agentId] || 1.0;

        // Separate decay rates for recovering vs degrading
        const isGood   = +(compliance > 0.5);
        const decay    = isGood * decayGood + (1 - isGood) * decayBad;
        const newHealth = current + decay * (compliance - current);

        this._health[agentId] = Math.max(0, Math.min(1, newHealth));

        // Log anomaly if health drops below threshold
        const isAnomalous = +(newHealth < 0.3);
        const entry = { tick: this.iteration, agentId, health: newHealth, compliance };
        this._anomalyLog.splice(0, 0, ...[entry].slice(0, isAnomalous));

        return newHealth;
    }

    getHealth(agentId) { return this._health[agentId] ?? 1.0; }

    // ─────────────────────────────────────────
    // AUTHORITY MASK — agents with low health
    // get reduced authority (soft isolation)
    // ─────────────────────────────────────────
    authorityMask(agentId, threshold = 0.3, sharpness = 10) {
        const h = this.getHealth(agentId);
        // Sigmoid: full authority above threshold, fades below
        return Semantic.softThresh(h, threshold, sharpness);
    }

    // ─────────────────────────────────────────
    // QUARANTINE ZONE — agents in quarantine
    // lose all outbound authority but still exist
    // ─────────────────────────────────────────
    isQuarantined(agentId, threshold = 0.15) {
        return +(this.getHealth(agentId) < threshold);
    }

    // ─────────────────────────────────────────
    // SYSTEM IMMUNITY — collective health
    // Average health across all registered agents
    // ─────────────────────────────────────────
    updateImmunity() {
        const vals = Object.values(this._health);
        if (vals.length === 0) { this._immunity = 1.0; return 1.0; }
        this._immunity = vals.reduce((a, b) => a + b, 0) / vals.length;
        return this._immunity;
    }

    getImmunity() { return this._immunity; }

    // ─────────────────────────────────────────
    // RECOVERY PULSE — boost health of agents
    // triggered by high immunity (herd recovery)
    // ─────────────────────────────────────────
    recoveryPulse(boostAmount = 0.05) {
        const immunityBonus = this._immunity * boostAmount;
        Object.keys(this._health).forEach(id => {
            this._health[id] = Math.min(1.0, this._health[id] + immunityBonus);
        });
    }

    // ─────────────────────────────────────────
    // ANOMALY LOG — read & trim
    // ─────────────────────────────────────────
    getAnomalyLog(last = 10) {
        return this._anomalyLog.slice(0, last);
    }

    clearAnomalyLog() { this._anomalyLog = []; }
}

export { ComplianceHealing };
